File: day1/puzzle2.lua
local function main()
    local sum = 0
    local f = io.open("puzzle_input.txt", "r")
    if not f then
        os.exit(1)
    end
    local number_digits = { 1, 2, 3, 4, 5, 6, 7, 8, 9 }
    local word_digits = { one = 1, two = 2, three = 3, four = 4, five = 5, six = 6, seven = 7, eight = 8, nine = 9 }
    local digits = {}
    for _, v in ipairs(number_digits) do
        table.insert(digits, tostring(v))
    end
    for k, _ in pairs(word_digits) do
        table.insert(digits, k)
    end
    local line_no = 0
    for line in f:lines() do
        line_no = line_no + 1
        local first_index = #line + 1
        local first = nil
        local last_index = -1
        local last = nil
        for _, digit in pairs(digits) do
            local start_index = 0
            local end_index = 0
            while true do
                start_index, end_index = line:find(digit, start_index + 1)
                if not start_index then break end
                if start_index < first_index then
                    first_index = start_index
                    first = line:sub(start_index, end_index)
                end
                if start_index > last_index then
                    last_index = start_index
                    last = line:sub(start_index, end_index)
                end
            end
        end
        print(line_no, first, last)
        local first_digit = word_digits[first]
        if not first_digit then first_digit = tonumber(first) end
        local last_digit = word_digits[last]
        if not last_digit then last_digit = tonumber(last) end
        sum = sum + first_digit * 10 + last_digit
    end
    f:close()
    print("sum = ", sum)
end

main()

File: day1/puzzle2.py
def main():
    num_digits = [str(n) for n in range(1, 10)]
    word_digits = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"] 
    digits = num_digits + word_digits
    sum = 0
    with open("puzzle_input.txt", "rt") as f:
        for line in f:
            first_index = len(line)
            first = None
            last_index = -1
            last = None
            for digit in digits:
                index = line.find(digit)
                if index == -1:
                    continue
                if index < first_index:
                    first_index = index
                    first = digit
                index = line.rfind(digit)
                if index == -1:
                    continue
                if index > last_index:
                    last_index = index
                    last = digit
            num = 0
            if first in num_digits:
                num = int(first)
            else:
                num = word_digits.index(first) + 1
            num *= 10
            if last in num_digits:
                num += int(last)
            else:
                num += word_digits.index(last) + 1
            sum += num
    print(f"sum = {sum}")


if __name__ == "__main__":
    main()
